metadata/environment.go

```diff
@@ -281,37 +285,43 @@ func (m *manager) SetEnvironment(name string, desired *Environment) error {
 		//
 		// Move the directory
 		//
+
 		pathOld := appendToAbsPath(m.environmentsPath, name)
 		pathNew := appendToAbsPath(m.environmentsPath, desired.Name)
-		jsonnetPathOld := string(appendToAbsPath(pathOld, path.Base(name)+".jsonnet"))
-		jsonnetPathNew := string(appendToAbsPath(pathOld, path.Base(desired.Name)+".jsonnet"))
 		exists, err := afero.DirExists(m.appFS, string(pathNew))
 		if err != nil {
 			return err
 		}
+
 		if exists {
-			return fmt.Errorf("Failed to create environment, directory exists at path: %s", string(pathNew))
-		}
-		// Need to first create subdirectories that don't exist
-		intermediatePath := path.Dir(string(pathNew))
-		log.Debugf("Moving directory at path '%s' to '%s'", string(pathOld), string(pathNew))
-		err = m.appFS.MkdirAll(intermediatePath, defaultFolderPermissions)
-		if err != nil {
-			return err
-		}
-		// note: we have to move the jsonnet file first because of a bug with afero: spf13/afero:#141
-		log.Debugf("Renaming jsonnet file from '%s' to '%s'", path.Base(jsonnetPathOld), path.Base(jsonnetPathNew))
-		err = m.appFS.Rename(jsonnetPathOld, jsonnetPathNew)
-		if err != nil {
-			log.Debugf("Failed to move path '%s' to '%s", jsonnetPathOld, jsonnetPathNew)
-			return err
-		}
-		// finally, move the directory
-		err = m.appFS.Rename(string(pathOld), string(pathNew))
-		if err != nil {
-			log.Debugf("Failed to move path '%s' to '%s", string(pathOld), string(pathNew))
-			return err
+			// we know that the desired path is not an environment from
+			// the check earlier. This is an intermediate directory.
+			// We need to move the file contents.
+			m.mvEnvFileContents(pathOld, pathNew)
+		} else if filepath.HasPrefix(string(pathNew), string(pathOld)) {
+			// the new directory is a children of the old directory --
+			// rename won't work.
+			err = m.appFS.MkdirAll(string(pathNew), defaultFolderPermissions)
+			if err != nil {
+				return err
+			}
+			m.mvEnvFileContents(pathOld, pathNew)
+		} else {
```

Comment:
A minor thing we should know about going in is that this allows us to have both an environment `us-west` and `us-west/dev`. I think this kinda sucks, but I don't feel super strongly about it.

Reply:
It's intentional -- agree it might not be *great* choice but I think this isn't a limitation we should impose on the user, if they really wanted to do that for some reason.

Reply:
I thought the plan was to make this a hierarchy. Otherwise `environments` could just be a flat directory of directories like `us-west_dev`. Let's file a bug and discuss: https://github.com/ksonnet/ksonnet/issues/138